Bug 1415868 - create hooks and roles for actions as hooks

djmitche opened the pull request.

djmitche committed:
Bug 1415868 - rename to make-scm-group-roles and do all levels

djmitche committed:
Bug 1415868 - add new action roles but leave the repo roles

djmitche committed:
Bug 1415868 - manage hook access in other mozilla-group roles

djmitche committed:
Bug 1415868 - read actions from a list of actions

djmitche committed:
Bug 1415868 - make a generic editHook function

djmitche committed:
Bug 1415868 - make push-specific roles and grant in-tree:.. to them

tomprince commented on src/make-scm-group-roles.js:
```diff
@@ -21,10 +21,25 @@ module.exports.run = async (options) => {
     var roleId = `mozilla-group:active_scm_level_${level}`;
-    var scopes = projectsWithGroup.map(project => {
+
+    // See https://bugzilla.mozilla.org/show_bug.cgi?id=1415868 for "old" and "new"
```
:frowning_face:

djmitche replied:
yeah, we need to not take scopes away from people until the new stuff works!

tomprince replied:
I'm more unhappy with the variables names. But if this exists for a short enough period, it won't be a problem.

tomprince commented on src/update-action-hook-perms.js:
```diff
@@ -0,0 +1,52 @@
+    .description('update action-related `hooks:trigger-hook:` scopes in mozilla-group:* roles');
+};
+
+EXPECTED_PERMS = [
```
I wonder if this wants to be in `ci-configuration`?

djmitche replied:
Yes, almost all of this does -- but I'm holding that for a follow-up.

tomprince commented on src/update-action-hook-perms.js:
```diff
@@ -0,0 +1,52 @@
+    {trustDomain: 'gecko', level: '3', actionPerm: 'relpromo'},
+  ]},
+  {group: 'releng', actions: [
+    {trustDomain: 'gecko', level: '3', actionPerm: 'relpromo'},
```
Probably want to give `comm`/`relpromo` here too.

Also, at least `releng` (but potentially wider) will want level 1/2 scopes for this too. We eventually want to be able to do staging release on try, and I've already been doing it a bit.

djmitche replied:
Oh, I meant to change this from relpromo to something more innocuous like purge-cache to start with.  I don't want to mess with "scary things" yet!  But yes, this will be the place to add such scopes.

tomprince commented on src/util/tcyml.js:
```diff
@@ -0,0 +1,12 @@
+
+const CENTRAL_RAW = 'https://hg.mozilla.org/mozilla-central/raw-file/default/';
+
+exports.getCentralTaskclusterYml = async () => {
```
I think it would make more sense for this to grabbed per-project (even if you only use it for m-c here).

djmitche replied:
We don't have a hook per project, though -- only per level.  We could potentially have one per repo, but then we have a hook per repo, per actionPerm, and we are granting those individual trigger-hook scopes to users, so it's quite an explosion.  I don't see a great benefit, either -- it would just leave us running `tcadmin` on every uplift, if there was something important changing in `.taskcluster.yml`.

tomprince replied:
I'm happy if you call it only with `mozilla-central` (although I think we also need to call it with `comm-central`), but I think it is worth making the utility general, even if the callers aren't).

tomprince commented on src/util/action-hooks.js:
```diff
@@ -0,0 +1,17 @@
+exports.ACTION_HOOKS = [
```
ci-configuration?

djmitche replied:
yup

tomprince commented on src/make-gecko-action-hooks.js:
```diff
@@ -0,0 +1,155 @@
+      tasks_for: 'action',
+      action: {
+        $merge: [
+          {$eval: 'payload.decision.action'},
```
Would there be any value in changing this so that the stuff that can be overridden is in a sub-key here, instead of merging? This will require some `.taskcluster.yml` refactoring, but I think it will be less error prone.

djmitche replied:
There are only two, and one of them (cb_name) is only sometimes overridden.  So I'm not sure that'd be less confusing.  Also, a sub-key would still require merging..

tomprince replied:
We could have (in python python psuedo-code):

```python
repo_scope = {...}
if action.actionPerm == 'generic':
   cb_name = action.pop('cb_name')
else:
    cb_name = action.actionPerm

task = {
    ...
    'action': {
        'repo_scope': repo_scope,
        'cb_name': {'$eval': 'payload.action.cb_name',
        'params': {'$eval': 'payload.action.params'},
    },
    ...
}
```
with bikesheading over `param`.
```

djmitche replied:
I laid out the three contexts concerned here in [this gist](https://gist.github.com/djmitche/b338559f8e1eae35e3e36a30f00759ed#file-contexts-md).

My initial reaction to your comment is that `stuffWeWillNotOverride` would be a better name than `params`.  It seems to be designing the data structure to suit the software constructing it (and I'm not even convinced the proposed design suits any better).  We could potentially flatten out all of the stuff in the `action` subkey.

I'd like to introduce minimal churn here, and keep risk of breaking existing (kind=task) actions as low as possible.  Is that risk worthwhile?

tomprince replied:
My concern is less about what the code looks like than about security; making it harder to future changes to accidentally let the security sensitive values be overridden (which suggests the name should be something like `untrusted`, maybe).

To be clear, I don't think the code as-is is insecure. I just think that it would be easy for somebody to not realize the security impact of the code (for example, that the order of merging is critical). By moving the security critical values outside of the user-provided dictionary, I think that makes mistakes here less likely.

djmitche replied:
Hm, I see.  So, a few things to change:
 * add detail to hooks' `triggerSchema` to indiciate specific keys in action, push, and repository
 * don't include `repo_scope` in the hookPayload (and don't generate it in the in-tree code)
 * Rework these overrides to be a little clearer that we either take the given value, or force it, and why (with some comments).

tomprince reviewed (changes requested):
This looks good. I've suggested a few changes, but this is involved enough that I think another pass after they are made is probably worthwhile.

djmitche committed:
Bug 1415868 - make getTaskclusterYml project-agnostic

djmitche committed:
Bug 1415868 - make diff display better, with context and no stray newlines

djmitche committed:
Bug 1415868 - add much more detail to triggerSchema

djmitche committed:
Bug 1415868 - specify each part of the 'action' property directly

Also, simplify some logic by requiring that repository.url not have a
trailing slash

djmitche committed:
Bug 1415868 - drop unnecessary repo_scope

djmitche committed:
Bug 1415868 - drop unnecessary task property

djmitche committed:
Bug 1415868 - use the taskId value from hooks for ownTaskId

djmitche committed:
Bug 1415868 - use comm-central's .taskcluster.yml for comm-central hooks

djmitche commented:
OK, this is substantially updated to reflect reviews and conversation on the topic.

tomprince commented on src/make-gecko-action-hooks.js:
```diff
@@ -66,33 +66,22 @@ module.exports.run = async function(options) {
+        taskGroupId: '${payload.decision.action.taskGroupId}',
+        // repo_scope is calculated in the hook's JSON-e rendering; note that this is limited
+        // by the hook's `hook-id:..` role, too
+        repo_scope: 'assume:repo:${payload.decision.repository.url[8:]}:action:' + action.actionPerm,
```
It took me a few minute to convince myself that this was safe. I think it would make sense to specifically call out that this can only allow scopes that are also given to the `hook-id:*` role, so that the validation of the repository url happens there.

djmitche replied:
Thanks - updated

tomprince replied:
:+1:

tomprince reviewed (approved):
This looks good! Thanks for the changes. I've just got a minor nit about an unclear comment.

jonasfj commented on src/util/action-hooks.js:
```diff
@@ -0,0 +1,24 @@
+    trustDomain: 'gecko',
+    level: '1',
+    actionPerm: 'purge-caches',
+    groups: ['taskcluster', 'vpn_sheriff'],
```
I see no reason this should be so limited :)

But it's a good example nonetheless...

djmitche replied:
Yeah, we can work out the permissions later -- this is almost just placeholder stuff right now.
